src/partition/repart.c

```diff
@@ -3557,6 +3589,149 @@ static int context_verity_hash(Context *context) {
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        rb = BIO_new_mem_buf(roothash, -1);
+        if (!rb)
+                return log_oom();
+
+        p7 = PKCS7_sign(cert, pk, NULL, rb, PKCS7_DETACHED|PKCS7_NOCERTS|PKCS7_NOATTR|PKCS7_BINARY);
+        if (!p7)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to calculate PKCS7 signature: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        pb = BIO_new(BIO_s_mem());
+        if (!pb)
+                return log_oom();
+
+        if (i2d_PKCS7_bio(pb, p7) < 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to retrieve PKCS7 signature: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (BIO_get_mem_ptr(pb, &bm) < 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to retrieve memory buffer: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        sig = memdup(bm->data, bm->length);
+        if (!sig)
+                return log_oom();
+
+        *ret_signature = TAKE_PTR(sig);
+        *ret_signature_size = bm->length;
+
+        return 0;
+#else
+        return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "openssl is not supported, cannot setup verity signature: %m");
+#endif
+}
+
+static int context_verity_sig(Context *context) {
+        int fd = -1, r;
+
+        assert(context);
+
+        LIST_FOREACH(partitions, p, context->partitions) {
+                _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+                _cleanup_free_ uint8_t *sig = NULL;
+                _cleanup_free_ char *text = NULL;
+                Partition *hp;
+                size_t sigsz, padsz;
+
+                if (p->dropped)
+                        continue;
+
+                if (PARTITION_EXISTS(p))
+                        continue;
+
+                if (p->verity != VERITY_SIG)
+                        continue;
+
+                assert_se(hp = p->siblings[VERITY_HASH]);
+                assert(!hp->dropped);
+
+                assert(arg_certificate);
+                assert(arg_certificate_size > 0);
+
+                if (fd < 0)
+                        assert_se((fd = fdisk_get_devfd(context->fdisk_context)) >= 0);
+
+                r = sign_verity_roothash(hp->roothash, &sig, &sigsz);
+                if (r < 0)
+                        return r;
+
+                r = json_build(&v,
+                               JSON_BUILD_OBJECT(
+                                        JSON_BUILD_PAIR("rootHash", JSON_BUILD_STRING(hp->roothash)),
+                                        JSON_BUILD_PAIR(
+                                                "certificateFingerprint",
+                                                JSON_BUILD_HEX(
+                                                        SHA256_DIRECT(arg_certificate, arg_certificate_size),
```

Comment:
hmm, i am pretty sure this is actually supposed to be calculated of the DER encoding, but this will be PEM

Reply:
The spec doesn't say anything about that, we should probably specify that